File: src/main/java/com/mpe85/grampa/matcher/impl/ActionRule.java
package com.mpe85.grampa.matcher.impl;

import com.google.common.base.Preconditions;
import com.mpe85.grampa.matcher.Action;
import com.mpe85.grampa.matcher.RuleContext;

public class ActionRule<T> extends AbstractRule<T> {
	
	public ActionRule(final Action<T> action) {
		this.action = Preconditions.checkNotNull(action, "An 'action' must not be null.");
	}
	
	@Override
	public boolean match(final RuleContext<T> context) {
		context.getValueStack().takeSnapshot();
		final boolean matched = action.run(context);
		context.getValueStack().removeSnapshot(!matched);
		return matched;
	}
	
	private final Action<T> action;
	
}

File: src/main/kotlin/com/github/mpe85/grampa/rule/ActionRule.kt
package com.github.mpe85.grampa.rule

import com.github.mpe85.grampa.context.ParserContext
import com.github.mpe85.grampa.context.RuleContext
import com.github.mpe85.grampa.util.checkEquality
import com.github.mpe85.grampa.util.stringify
import java.util.Objects.hash

/**
 * An action rule implementation.
 *
 * @param[T] The type of the stack elements
 * @param[action] The action that is executed by the rule
 * @param[skippable] Whether the action is skippable inside test rules
 * @author mpe85
 */
internal class ActionRule<T>(
    private val action: (RuleContext<T>) -> Boolean,
    private val skippable: Boolean = false,
) : AbstractRule<T>() {

    override fun match(context: ParserContext<T>): Boolean =
        context.inTestRule && skippable || action(context)

    override fun hashCode(): Int = hash(super.hashCode(), action, skippable)

    override fun equals(other: Any?): Boolean =
        checkEquality(other, { super.equals(other) }, { it.action }, { it.skippable })

    override fun toString(): String = stringify("action" to action, "skippable" to skippable)
}
